### DISCUSSION TITLE
Touch on Sunton ESP32-8048S043C

### DISCUSSION THREAD

**Discussion by oschettler** [HASwitchPlate/openHASP#574] [Q&A] [answered]
I've installed sunton-8048s043c_full_16MB_v0.7.0-rc6_2f4092b.bin 
The board has an 800x480 capacitive touch screen. 

The display works nicely. However, I've noticed that the touch interface is of an incorrect scale:

![SCR-20231112-lkjy-2](https://github.com/HASwitchPlate/openHASP/assets/124634/2652d847-748b-4700-91d0-0ebd0421e2f2)

When I touch the display a little below and right of the center, the pointer already is in the lower right corner. 

The configuration for this board starts here: https://github.com/HASwitchPlate/openHASP/blob/6cf4262f2356eaa686abe045ab01a5467a6963b1/user_setups/esp32s3/sunton-esp32-s3-tft.ini#L123

I see TFT_WIDTH=800 and TFT_HEIGHT=480, but also TOUCH_WIDTH=480 and TOUCH_HEIGHT=272. Maybe this is the problem? 

Best,
-- Olav

**Comment by fvanroie** [comment]
Are you sure it is the ESP32-8048S043C and not the ESP32-8048S050C ? Which link did you buy the display from?

If it is an actual ESP32-8048S043C (check the PCB code on the backside), maybe they changed the hardware revision of the touchscreen...

**Reply by oschettler** [reply]
The board has "ESP32-8048S043" on the back.

**Comment by oschettler** [comment, marked as answer]
I have compiled a version with TOUCH_WIDTH and THOUCH_HEIGHT fixed:

````
diff --git a/user_setups/esp32s3/sunton-esp32-s3-tft.ini b/user_setups/esp32s3/sunton-esp32-s3-tft.ini
index 7cbb14d8..73eec22e 100644
--- a/user_setups/esp32s3/sunton-esp32-s3-tft.ini
+++ b/user_setups/esp32s3/sunton-esp32-s3-tft.ini
@@ -144,8 +144,8 @@ build_flags =
     -D TFT_PREFER_SPEED=14000000  ; 1/2 of Typical DCLK Frequency
     -D TFT_AUTO_FLUSH=1
     ; Touch Settings
-    -D TOUCH_WIDTH=480
-    -D TOUCH_HEIGHT=272
+    -D TOUCH_WIDTH=800
+    -D TOUCH_HEIGHT=480
     -D TOUCH_DRIVER=0x911
     -D TOUCH_SCL=20
     -D TOUCH_SDA=19
````

To make it compile, in platform_override.ini, I also had to add: 

````
extra_default_envs =
    sunton-8048s043c_16MB
````

I also needed to comment the upload_speed in platform.ini:

````
[env]
;upload_speed = 921600
````

With this, I get the full screen width for touch.

Thank you for this fine software.
-- Olav

**Comment by Gemelon** [comment]
I have the exact same display and I have the exact same problem, how do you resolve the problem?

**Comment by RyanEwen** [comment]
Do you guys get backlight flicker at 100% and a rolling effect at lower brightnesses? I also get some artifacts when changing pages. Not sure if it's just me

**Reply by RyanEwen** [reply]
Following-up, I switched to ESPHome with the new LVGL widget support and have working configs for a handful of devices here: https://github.com/RyanEwen/esphome-lvgl/tree/main/devices

The key thing with the larger CYDs was to use specific older versions of `esp-idf`. Maybe that can help someone here.

**Comment by Rolibaer** [comment]
Test Report – Display ESP32-8048S043C with RC12 & RC15

    Hardware: ESP32-8048S043C
    Touch Controller: GT911
    Firmware Versions Tested: RC12, RC15
    Build Environment: Gitpod (compiled .bin file with platform_override.ini from oschettler)
    Web Interface: Working as expected
    Display: Not working (white screen...)
    Touch (GT911): Not responding

Issue Description:
After flashing RC12 and RC15 firmware to the ESP32-8048S043C, the web interface is fully functional. However, the display remains blank and shows no content. The touch controller (GT911) is also unresponsive.

Questions:

    Have there been any recent changes to display/touch drivers between RC6 and RC15?
    Is the ESP32-8048S043C supported in these release candidates?

**Reply by Rolibaer** [reply]
in between i got it working **but not the Touch**....
My config is (sunton-esp32-s3-tft.ini):
```
 [env: sunton-8048s043c_16MB]
extends = sunton-esp32-s3-tft, flash_16mb
```

```
build_flags =
    -D HASP_MODEL="Sunton ESP32-8048S043C"
    ${sunton-esp32-s3-tft.build_flags}
    ${sunton-tft-common-pins.build_flags}
    ; Bus Settings
    -D LV_VDB_SIZE=76800 ; 10% of full framebuffer
    -D TFT_WIDTH=800
    -D TFT_HEIGHT=480
    ; Panel Settings
    -D TFT_HSYNC_POLARITY=0
    -D TFT_HSYNC_FRONT_PORCH=8   ; Maximum HSYNC Front Porch
    -D TFT_HSYNC_PULSE_WIDTH=4    ; Typical HSYNC Pulse Width
    -D TFT_HSYNC_BACK_PORCH=8     ; Typical HSYNC Back Porch
    -D TFT_VSYNC_POLARITY=0
    -D TFT_VSYNC_FRONT_PORCH=8   ; Maximum VSYNC Front Porch
    -D TFT_VSYNC_PULSE_WIDTH=4    ; Typical VSYNC Pulse Width
    -D TFT_VSYNC_BACK_PORCH=8     ; Typical VSYNC Back Porch
    -D TFT_PCLK_ACTIVE_NEG=1
    -D TFT_PREFER_SPEED=14500000  ; 1/2 of Typical DCLK Frequency
    -D TFT_AUTO_FLUSH=1
    ; Touch Settings
    -D TOUCH_WIDTH=800                                                                           <<<<<<<---------
    -D TOUCH_HEIGHT=480                                                                          <<<<<<<---------
    -D TOUCH_DRIVER=0x911
    -D TOUCH_SCL=20
    -D TOUCH_SDA=19
    -D TOUCH_IRQ=-1
    -D TOUCH_RST=38
    -D I2C_TOUCH_ADDRESS=0x5d ; or 0x5D or 0x14
    -D I2C_TOUCH_FREQUENCY=400000
lib_deps =
    ${sunton-esp32-s3-tft.lib_deps}
    ${goodix.lib_deps}
```


On the debugport i get:
```
 GUI : Initialize Cursor
DRVR: GT911 Gestartet (480x272)                                                             <<<<<<<---------
DRVR:     *  Found device 0x5d
```
 
compiled with GitPod
```

platformio_override.ini: 
 extra_default_envs = sunton-8048s043c_16MB
```

platformio.ini: 
```
[env]
 ; upload_speed = 921600
```


has anybody an IDEA?

**Reply by Rolibaer** [reply]
In the meantime I have found a solution:
I think it was a wrong configuration on the GT911
After testing a TouchPrint.ino file it worked. (I think it resets the GT911, i did not do any calibration)
```
#include` "TAMC_GT911.h"

#define TOUCH_SDA  19
#define TOUCH_SCL  20
#define TOUCH_INT -1
#define TOUCH_RST 38
#define TOUCH_WIDTH  800
#define TOUCH_HEIGHT 480

TAMC_GT911 tp = TAMC_GT911(TOUCH_SDA, TOUCH_SCL, TOUCH_INT, TOUCH_RST, TOUCH_WIDTH, TOUCH_HEIGHT);

void setup() {
  Serial.begin(115200);
  Serial.println("TAMC_GT911 Example: Ready");
  tp.begin();
  tp.setRotation(ROTATION_NORMAL);
}

void loop() {
  tp.read();
  if (tp.isTouched){
    for (int i=0; i<tp.touches; i++){
      Serial.print("Touch ");Serial.print(i+1);Serial.print(": ");;
      Serial.print("  x: ");Serial.print(tp.points[i].x);
      Serial.print("  y: ");Serial.print(tp.points[i].y);
      Serial.print("  size: ");Serial.println(tp.points[i].size);
      Serial.println(' ');
    }
  }
}
```

**Reply by Rolibaer** [reply]
just pulled a BUG #896 for the display config

**Reply by ELNL** [reply]
I found out following. I have 2 versions of the module: 8048S043C and 8048S043C_I.

Both work fine but there is a difference between the firmware of the touch controller. I have version 65 and 66 (decimal). This can be determent by reading register 0x8047 op the GT911 chip. Version 66 is found on the 8048C043C module.

If the version is 66, then the touch chip outputs (0,0) to (800x480) about.
If the version is 65, then the touch chip outputs (0,0) to (480x280) about.

It is strange that manufacterer of the C_I variant mounts a chip with configuration not belonging to the displaytype. However with this knowledge one can make the software figure out the needed callibration for every known version. So in my case both hardware versions run on the same software.

**Reply by fvanroie** [reply]
Thanks @ELNL , that's very useful information indeed. I have reworked the GT911 driver to use the maximum coordinates from the firmware registers. Can you please test with this firmware below on both devices and see if it auto-detects the touch resolution?

https://github.com/HASwitchPlate/openHASP/actions/runs/29157291451

`TOUCH_WIDTH` and `THOUCH_HEIGHT` are now optional and not needed for the GT911 driver to know the maximum coordinates.

**Reply by ELNL** [reply]
I looked at the link. Important to know, Once more learning, I am not a software develloper. Also, only experienced with the ESP-IDF platform, not Arduino.
But, the idea to read the max coordinates out of registers sonds likea great option. The link points to code I cannot deploy or fully understand. Could you point out how the members xResolution and yResolution from 'TouchGt911* instance" get the value?

I'll try reading the values of registers 0x8048 .. 0x804B of both modules. Maybe they come up with the right values. It is late now, It will take 1 or 2 days for me to do the test . Grtzz.

**Reply by fvanroie** [reply]
The code is here, it already reads the info block which includes the resolution: https://github.com/HASwitchPlate/openHASP/blob/118fc6e1bd8dcd0e8c26dc5d3a9c4569856d45df/src/drv/touch/touch_driver_gt911.cpp#L88-L99

You do not need to change this yourself or compile it yourself, you can download Sunton.zip from the bottom of the Build Actions, under Artifacts:
https://github.com/HASwitchPlate/openHASP/actions/runs/29157291451

The ESP32-8048S043C firmware bin files are already compiled in the Sunton.zip file.
You should see in the serial log a line that says `GT911 Service started (800x480)` or `GT911 Service started (480x272)`.
That line should already appear in your current firmware too. Just check the serial log at (re)boot.

**Reply by ELNL** [reply]
I am sorry I don't know how to download code into the target. I use the Visual Code IDE with the plugin from Espressif. The workflow is just execute some pyton call and the flash process runs. For me this proces is a 'black box'.

But, maybe I can be of some help. I made two functions that read following of the chip:
<img width="321" height="275" alt="image" src="https://github.com/user-attachments/assets/31e0a523-3f9c-491b-ad7a-7d2a913067dd" />

```
bool readTouchChip8bitRegister(const uint16_t reg, uint8_t * value)
{
    if (touchInstanceHandle == NULL)
    {
        return false;
    }
    if (touchInstanceHandle->io == NULL)
    {
        return false;
    }
    if (esp_lcd_panel_io_rx_param(touchInstanceHandle->io, reg, value, sizeof(uint8_t)) != ESP_OK)
    {
        return false;
    }
    return true;
}

bool readTouchChip16bitRegister(const uint16_t reg, uint16_t * value)
{
    uint8_t lowByte = 0;
    if (!readTouchChip8bitRegister(reg, &lowByte))
    {
        return false;
    }
    uint8_t hiByte = 0;
    if (!readTouchChip8bitRegister(reg + 1, &hiByte))
    {
        return false;
    }
    *value = (((uint16_t)hiByte) << 8) + (uint16_t)lowByte;
    return true;
}
```

After initialisation of the GT911 chip I added this code to read the maximum values the GT911 chip would give. With these numbers I can scale the position of the touch to send to LVGL.

```
#define ESP_GT911_REGISTER_CONFIG_VERSION   0x8047
#define ESP_GT911_REGISTER_OUT_MAX_X_LOW    0x8048
#define ESP_GT911_REGISTER_OUT_MAX_X_HI     0x8049
#define ESP_GT911_REGISTER_OUT_MAX_Y_LOW    0x804A
#define ESP_GT911_REGISTER_OUT_MAX_Y_HI     0x804B

```

```
    uint8_t versionByte;
    if (esp_lcd_panel_io_rx_param(touchInstanceHandle->io, ESP_GT911_REGISTER_CONFIG_VERSION, &versionByte, 1) != ESP_OK)
    {
        ESP_LOGE(TAG, "%s: fail reading version", __func__);
        return false;
    }
    if (!readTouchChip16bitRegister(ESP_GT911_REGISTER_OUT_MAX_X_LOW, &touchHres))
    {
        ESP_LOGE(TAG, "%s: fail Hres", __func__);
        return false;
    }
    if (!readTouchChip16bitRegister(ESP_GT911_REGISTER_OUT_MAX_Y_LOW, &touchVres))
    {
        ESP_LOGE(TAG, "%s: fail Vres", __func__);
        return false;
    }

    ESP_LOGI(TAG, "%s: current version=%u Hres:%u Vres:%u", __func__, versionByte, touchHres, touchVres);
    return true;

```

This works for both display versions 65 and 66 perfect. This eliminates constants in  the code.

Version 65 outputs:
`I (1194) Display: Display8048S043_Touch_init: current version=65 Hres:480 Vres:272
`
Version 66 outputs:
`I (1194) Display: Display8048S043_Touch_init: current version=66 Hres:800 Vres:480`

Hope you can use this confirmation.

**Reply by ELNL** [reply]
The performance of the two different displays is not noticable.

https://github.com/user-attachments/assets/289e462b-0b0e-4e35-98b2-58fa097db989
